### DISCUSSION TITLE
Default-network-driver & Enterprise-wide message model

### DISCUSSION THREAD

**Discussion by whalezup** [syslog-ng/syslog-ng#4085] [Q&A] [answered]
[Default-network-driver & Enterprise-wide message model]

This is a use case question and I am new to syslog-ng. Appreciate any advice from the anyone here. 

Currently I am trying to structure a workable approach for logs ingestion from varying sources into a central log server via a relay server. 

(A) >> (B) >> (C)  >> (D)

where
(A) = LSH (network devices/winOS/linux)  
(B) = Relay server (Syslog-ng PE or Syslog-ng OSE)  
(C) = Central (Syslog-ng store box)
(D)	= SIEM 

* Concept is to apply default-network-driver + EWMM at (B) Relay 

Rational for concept: As I have read from the admin guide, this default network driver can "parse several different types of syslog messages from the network" , "this source tries to parse the messages automatically" , "sets the ${.app.name} name-value pair to the name of the application that sent the log message" 

For EWMM, I also read "It does not matter if you parse the messages on the client, on a relay, or on the central server, their structured results will be available where you store the messages."

All above seems like very useful features to apply. 

* Example default-network-driver config

source sdnd {
    default-network-drivers(
       rfc5424-tcp-port(602)
       rfc5424-tls-port(6515)
       tcp-port(601)
       udp-port(514)
       flags(sanitize-utf8)
    );
};

* (B) will send to (C)'s source port 602 configured with syslog() driver
* (C) to (D) still not thought of.

[Questions]
1) Default-network-drivers has a max-connection default of 10 (per port). How high can it be adjust it to? What determines the limits of this value ?

2) In EWMM codes, i saw a note about it being "experimental transport for transferring messages in whole between syslog-ng instances". Does it mean, it is still not meant for production usage?

I appreciate any advice with example configuration statement if above is a viable approach. Thank you!

**Comment by alltilla** [comment]
Hi @whalezup 

Answering you questions:
  1. Technically `max-connection()` can be set up to int32 max (2147483647), but that is not advised. There is an option called `log-fifo-size()` which limits the number of messages stored in memory for each source. When this limit is reached either messages will be dropped, or reading from the source will be paused, depending on how `flow-control` was set. In case of `tcp` based sources the value of `log-fifo-size()` is distributed evenly between the number of possible connections. E.g.: `log-fifo-size(10000)` and `max-connections(1000)` will only store 10 messages in memory at a given time for any of the connections. This is true, even if only 1 client is connected, because `syslog-ng` has to reserve space for the possible other clients. This problem can be solved with [dynamic flow-control](https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.37/administration-guide/61#TOPIC-1829144), but it is only available in OSE, it is not supported in PE. In a nutshell `max-connections()` should be set to a reasonable value related to the expected number of connections, and a sufficient `log-iw-size()` value should be set accordingly.
  2. Although there is a comment about EWMM being experimental, it is documented in the [syslog-ng PE Admin Guide](https://support.oneidentity.com/technical-documents/syslog-ng-premium-edition/7.0.29/administration-guide/60#TOPIC-1740281), so it is supported by One Identity to use in syslog-ng PE.

Please note that the GitHub communication platforms are for solely OSE users, help here is only provided by community members in their free time or by syslog-ng developers with a limited availability. If you are a syslog-ng PE customer, please contact [One Identity support](https://support.oneidentity.com/contact-support/) with any of your syslog-ng PE or SSB related questions. Thank you!

Cheers,
Attila

**Reply by whalezup** [reply]
@alltilla 
Much appreciated your prompt & detail inputs and advices. It is useful.
I will also be mindful on posting further queries related to syslog-ng PE / SSB here. :)

**Reply by alltilla** [reply]
Thanks! Happy to help :)

**Reply by MrAnno** [reply]
Just to be precise, our queueing mechanism works differently when flow-control is enabled/disabled.

On a flow-control enabled path, `log-iw-size()` will be divided evenly for `max-connection()` separate connections. That value will be the max queue size of each connection. `log-fifo-size()` is not relevant on flow-controlled log paths.

On non-flow-controlled paths, the destination's `log-fifo-size()` option will be honored. That queue is "shared", there will be no even distribution across connections (first-come, first-served). `log-iw-size()` is not really relevant on non-flow-controlled paths, it can only be used for performance tuning.

**Comment by bazsi** [comment, marked as answer]
Hi,

Thanks for your questions, please find my answers inline.


On Wed, Jul 20, 2022 at 8:56 AM whalezup ***@***.***> wrote:

> [Default-network-driver & Enterprise-wide message model]
>
> This is a use case question and I am new to syslog-ng. Appreciate any
> advice from the anyone here.
>
> Currently I am trying to structure a workable approach for logs ingestion
> from varying sources into a central log server via a relay server.
>
> (A) >> (B) >> (C) >> (D)
>
> where
> (A) = LSH (network devices/winOS/linux)
> (B) = Relay server (Syslog-ng PE or Syslog-ng OSE)
> (C) = Central (Syslog-ng store box)
> (D) = SIEM
>

This architecture is a good one, Relays can be uses in branch
offices/locations so that they take on the logs from a specific location
and make sure they make it into the central, with disk buffering if needed.

Before reading any further, please note that I am not related to One
Identity anymore (but I am with the open source project), the commercial
entity behind syslog-ng, so my information on this may not be completely
accurate. Ask your local sales team about these questions.

Last I checked, the syslog-ng store box didn't have support for
EWMM/default-network-drivers, so you might want to put a "central relay"
receives logs from branch offices, and which ships to both SSB and your
SIEM. Stock syslog-ng (be that OSE or PE) moves faster in this direction.

Also, syslog-ng relay in the central, can be adjusted, customized, scaled
easier than SSB, which is a closed appliance, with limitations. But your
milage may vary.

If you meant that (B) is running in the central already, then this
architecture is fine.


>
>    - Concept is to apply default-network-driver + EWMM at (B) Relay
>
> Rational for concept: As I have read from the admin guide, this default
> network driver can "parse several different types of syslog messages from
> the network" , "this source tries to parse the messages automatically" ,
> "sets the ${.app.name} name-value pair to the name of the application
> that sent the log message"
>
> For EWMM, I also read "It does not matter if you parse the messages on the
> client, on a relay, or on the central server, their structured results will
> be available where you store the messages."
>
> All above seems like very useful features to apply.
>
>    - Example default-network-driver config
>
> source sdnd {
> default-network-drivers(
> rfc5424-tcp-port(602)
> rfc5424-tls-port(6515)
> tcp-port(601)
> udp-port(514)
> flags(sanitize-utf8)
> );
> };
>

looks good to me. BTW: you can always look at what this translates to,
here:
https://github.com/syslog-ng/syslog-ng/blob/master/scl/default-network-drivers/plugin.conf


>
>    - (B) will send to (C)'s source port 602 configured with syslog()
>    driver
>
>



>
>    - (C) to (D) still not thought of.
>
> [Questions]
>
>    1.
>
>    Default-network-drivers has a max-connection default of 10 (per port).
>    How high can it be adjust it to? What determines the limits of this value ?
>
>
default-network-drivers() has a max-connections() parameter that you can
customize.


>
>    1.
>    2.
>
>    In EWMM codes, i saw a note about it being "experimental transport for
>    transferring messages in whole between syslog-ng instances". Does it mean,
>    it is still not meant for production usage?
>
> I appreciate any advice with example configuration statement if above is a
> viable approach. Thank you!
>

ewmm() is just a message format that allows both sending and receiving any
name-value pairs. I would consider this production ready, as I don't expect
changes in the format now. But even if it changes, a local transition is
possible.


Bazsi

**Reply by whalezup** [reply]
@bazsi 
Thank you for addressing my queries and sharing of your insights.
I have taken note on your "central relay" concept and will work towards that direction.
